src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+        } else {
+            return op;
+        }
+    }
+
+public:
+    SubstituteSomeLets(size_t _count)
+        : count(_count) {
+    }
+};
+
+void make_const_interval(Interval &interval) {
+    // Note that we can get non-const but well-defined results (e.g. signed_integer_overflow);
+    // for our purposes here, treat anything non-const as no-bound.
+    if (!is_const(interval.min)) {
+        interval.min = Interval::neg_inf();
+    }
+    if (!is_const(interval.max)) {
+        interval.max = Interval::pos_inf();
+    }
+}
+
+Expr push_rationals(const Expr &expr, const Direction direction) {
+    if (expr.type() == Int(32)) {
+        return handle_push_none(expr, direction);
+    } else {
+        return expr;
+    }
+}
+
+std::pair<Expr, bool> remove_unbounded_terms(const Expr &expr, Direction direction, const Scope<Interval> &scope) {
+    std::map<std::string, int> var_uses;
+    CountVarUses counter(var_uses);
+    expr.accept(&counter);
+    // To strip unbounded/uncorrelated terms, we need to know use counts.
+    StripUnboundedTerms tool(direction, &scope, var_uses);
+    Expr mut = tool.mutate(expr);
+    return {mut, tool.unbounded_vars != 0};
+}
+
+Expr reorder_terms(const Expr &expr) {
+    return ReorderTerms().mutate(expr);
+}
+
+Expr substitute_some_lets(const Expr &expr) {
+    // Empirically appears to be a good value, most real
+    // expressions we see won't have more substitutions
+    // than this, but for some expressions, larger values
+    // can cause Expr size explosion.
+    static const size_t count = 16;
+    return SubstituteSomeLets(count).mutate(expr);
+}
+
+Expr approximate_optimizations(const Expr &expr, Direction direction, const Scope<Interval> &scope) {
+    Expr subs = substitute_some_lets(expr);
+    if (!subs.same_as(expr)) {
+        subs = simplify(subs);
+    }
+
+    auto [cleaned, unbounded] = remove_unbounded_terms(subs, direction, scope);
+
+    if (unbounded) {
+        return Expr();
+    } else if (!cleaned.same_as(subs)) {
+        cleaned = simplify(cleaned);
+    }
+
+    subs = reorder_terms(cleaned);
+    // TODO: only do push_rationals if correlated divisions exist.
+    Expr pushed = push_rationals(subs, direction);
+    if (!pushed.same_as(subs)) {
+        pushed = simplify(pushed);
+    }
+    return pushed;
+}
+
+bool possibly_correlated(const Expr &expr) {
+    // TODO: is there a better/cheaper way to detect possible correlations?
+    std::map<std::string, int> var_uses;
+    CountVarUses counter(var_uses);
+    expr.accept(&counter);
+    for (const auto &iter : var_uses) {
+        if (iter.second > 1) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool should_use_approximate_methods(const Expr &expr) {
+    return !is_const(expr) && expr.type() == Int(32) && possibly_correlated(expr);
+}
+
+Interval approximate_constant_bounds(const Expr &expr, const Scope<Interval> &scope) {
+    Interval interval;
```

Comment:
Would be more readable if you initialized this to everything or nothing. I don't remember what the default constructor does.

Reply:
fixed in cf591f763